base/test.jl

```diff
@@ -1087,4 +1167,142 @@ Base.convert(::Type{GenericString}, s::AbstractString) = GenericString(s)
 Base.endof(s::GenericString) = endof(s.string)
 Base.next(s::GenericString, i::Int) = next(s.string, i)
 
+function move_to_node1!(tests, node1, ts)
+    for t in ts
+        if t in tests
+            splice!(tests, findfirst(tests, t))
+            push!(node1, t)
+        end
+    end
+end
+
+function print_test_statistics(test, test_stats, worker, alignments)
+    name_str = rpad(test*" ($worker)", alignments[1], " ")
+    time_str = @sprintf("%7.2f", test_stats[2])
+    time_str = rpad(time_str, alignments[2], " ")
+    gc_str   = @sprintf("%5.2f", test_stats[5].total_time/10^9)
+    gc_str   = rpad(gc_str, alignments[3] ," ")
+    # since there may be quite a few digits in the percentage,
+    # the left-padding here is less to make sure everything fits
+    percent_str = @sprintf("%4.1f",100*test_stats[5].total_time/(10^9*test_stats[2]))
+    percent_str = rpad(percent_str,alignments[4]," ")
+    alloc_str = @sprintf("%5.2f",test_stats[3]/2^20)
+    alloc_str = rpad(alloc_str,alignments[5]," ")
+    rss_str = @sprintf("%5.2f",test_stats[6]/2^20)
+    rss_str = rpad(rss_str,alignments[6]," ")
+    print_with_color(:white, name_str, " | ", time_str, " | ", gc_str, " | ", percent_str, " | ", alloc_str, " | ", rss_str, "\n")
+end
+
+function runtests(names=["all"]; test_dir=joinpath(JULIA_HOME, Base.DATAROOTDIR, "julia/test/"), numcores::Int=ceil(Int, Sys.CPU_CORES / 2))
+    include(joinpath(test_dir, "choosetests.jl"))
```

Comment:
AFAICT doing this in `choosetests.jl` instead of `runtests.jl` doesn't really make it any easier for the packages. The most useful part of the base `choosetests.jl` is (IMHO) the `--skip` option and the `all` "meta"-test and this currently doesn't make it possible for the package to use it. It might be useful to provide a few helper functions for cmdline parsing but I honestly don't think the one we have in base is a particularly good one and packages are better off using sth like `ArgParse.jl` for that.

Even if a `choosetests` type of interface is wanted, it should be implemented as a callback rather than `include`ing a file.